GEODE-8782: Add getPrincipal method to FunctionContext interface

jdeppe-pivotal opened the pull request:
- Add the ability to retrieve the Principal from the FunctionContext
  when a SecurityManager is enabled.

Authored-by: Jens Deppe <jdeppe@vmware.com>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

jinmeiliao commented:
I see you are binding the subject to the threads running the function and also adding the the parameter `principal` to the constructor of the `FunctoinContext`, I wonder if we only need to do one of the two? If subject is already in the thread context, you don't need to pass it in the constructor, you can just directly get it from the thread local by calling `getCache().getSecurityService().getPrinicipal()`

upthewaterspout commented on geode-core/src/main/java/org/apache/geode/internal/cache/execute/FunctionRemoteContext.java:
```diff
@@ -84,6 +85,10 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
     }
     this.isReExecute = DataSerializer.readBoolean(in);
+
+    if (StaticSerialization.getVersionForDataStream(in).isNotOlderThan(KnownVersion.GEODE_1_14_0)) {
```
Instead of this, I think you could make this class implement VersionedDataSerializable which might handle writing the principal differently.

jdeppe-pivotal replied:
I tried that but had some upgrade tests that became flaky. I may have done something wrong, but using `VersionedDataSerializable` seemed to not work for this.

jdeppe-pivotal commented:
@jinmeiliao Yes, good idea. I've added the principal to `FunctionContextImpl` and removed the code that rebound the `Subject` on function threads.

@upthewaterspout I've also removed the default implementation for `FunctionContext.getPrincipal()` since FunctionContext (and impls) are never serialized anywhere and it is also never implemented by user code.

jinmeiliao commented:
After you are done with this, can you file another jira ticket to redo the `QueryCommand`? Currently we are sending the principal using `DataCommandRequest`, now that the principal is in the function context, we don't need to do that anymore. Or if you can lump it into this PR, that would be great too!
